exasol/ds/sandbox/runtime/ansible/roles/jupyter/files/notebook/cloud/02_s3_vs_reuters.ipynb

```diff
@@ -0,0 +1,317 @@
+    "* `id` (named as `NEWS_ID`): article identifier stored as a string\n",
+    "* `body`: text of the article up to 1024 characters long\n",
+    "\n",
+    "In addition to those 3 fields, we have field `topics`, which is extracted as a reference table.\n",
+    "In the json data topics are represented as a list of tags associated with the article, for example:\n",
+    "\n",
+    "```\n",
+    "    \"topics\": [\n",
+    "      \"grain\",\n",
+    "      \"wheat\",\n",
+    "      \"corn\",\n",
+    "      \"barley\",\n",
+    "      \"oat\",\n",
+    "      \"sorghum\"\n",
+    "    ],\n",
+    "```\n",
+    "\n",
+    "To have convenient access to topics, we use `toTableMapping` EDML feature, which creates a second table `NEWS_TOPICS`, keeping list of topics and `NEWS_ID`:\n",
+    "\n",
+    "```\n",
+    "      \"topics\": {\n",
+    "        \"toTableMapping\": {\n",
+    "          \"mapping\": {\n",
+    "            \"toVarcharMapping\": {\n",
+    "              \"destinationName\": \"NAME\"\n",
+    "            }\n",
+    "          }\n",
+    "        }\n",
+    "      },\n",
+    "```"
+   ],
+   "id": "5d450a20e2ce88b5"
+  },
+  {
+   "metadata": {},
+   "cell_type": "code",
+   "outputs": [],
+   "execution_count": null,
+   "source": [
+    "# put edml to bucket fs\n",
+    "bfs_bucket = open_bucketfs_connection(ai_lab_config)\n",
+    "bfs_path = bfs_utils.put_file(bfs_bucket, pathlib.Path(\"reuters-edml.json\"))"
+   ],
+   "id": "e46e42baa2a61831"
+  },
+  {
+   "metadata": {},
+   "cell_type": "code",
+   "outputs": [],
+   "execution_count": null,
+   "source": "bfs_path.as_udf_path()",
+   "id": "f0af003ba35b9ff0"
+  },
+  {
+   "metadata": {},
+   "cell_type": "code",
+   "outputs": [],
+   "execution_count": null,
+   "source": [
+    "sql = \"\"\"\n",
+    "CREATE VIRTUAL SCHEMA {schema_name!i}_VS USING {schema_name!i}.S3_FILES_ADAPTER WITH\n",
+    "    CONNECTION_NAME = 'S3_CONNECTION_VS'\n",
+    "    MAPPING         = {map_file!s};\n",
+    "\"\"\"\n",
+    "\n",
+    "with open_pyexasol_connection(ai_lab_config) as conn:\n",
+    "    conn.execute(sql, query_params={\n",
+    "        \"schema_name\": ai_lab_config.db_schema,\n",
+    "        \"map_file\": bfs_path.as_udf_path()\n",
+    "    })"
+   ],
+   "id": "e11cb124291c57ce"
+  },
+  {
+   "metadata": {},
+   "cell_type": "markdown",
+   "source": [
+    "# Query the data\n",
+    "\n",
+    "Once we have our Virtual Schema created, new table `AI_LAB_VS.NEWS` is available for query. \n",
+    "But you need to be aware of performance overhead. Virtal Schemas don't do any intermediate caching and every request to the Virtual Schema table performs several S3 data reads, which might be slow.\n",
+    "\n",
+    "If you're going to access the data several times, it might make sense to create the new table and copy the data from VS table in a normal table. Here we don't do that for simplicity - we do just one query and that's fine. "
```

Comment:
```suggestion
    "If you are going to access the data several times, it might make sense to create the new table and copy the data from the virtual table into a normal table. Here we don't do that for simplicity. "
```